File: solution/python3/141.linked-list-cycle.py
#
# @lc app=leetcode id=141 lang=python3
#
# [141] Linked List Cycle
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        runner = head
        walker = head
        while True:
            if runner is None or runner.next is None or runner.next.next is None:
                return False

            walker = walker.next
            runner = runner.next.next

            if walker == runner:
                return True


# @lc code=end

File: solution/java8/141.linked-list-cycle.java
/*
 * @lc app=leetcode id=141 lang=java
 *
 * [141] Linked List Cycle
 */

// @lc code=start
// Definition for singly-linked list.
// class ListNode {
//     int val;
//     ListNode next;
//     ListNode(int x) {
//         val = x;
//         next = null;
//     }
// }
public class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null) {
            return false;
        }
        ListNode runner = head;
        ListNode walker = head;

        while (true) {
            if (walker.next != null) {
                walker = walker.next;
            } else {
                return false;
            }

            if (runner.next != null && runner.next.next != null) {
                runner = runner.next.next;
            } else {
                return false;
            }

            if (walker.equals(runner)) {
                return true;
            }
        }

    }
}
// @lc code=end

